samples/nrf9160/nrf_cloud_coap_client/src/main.c

```diff
@@ -0,0 +1,803 @@
+			LOG_ERR("Timeout waiting for cells: %d", err);
+			err = 0;
+			if (!IS_ENABLED(CONFIG_WIFI)) {
+				break;
+			}
+		}
+
+		(void)k_mutex_lock(&cell_info_mutex, K_FOREVER);
+
+		if (cell_info.current_cell.id != LTE_LC_CELL_EUTRAN_ID_INVALID) {
+			LOG_INF("Current cell info: Cell ID: %u, TAC: %u, MCC: %d, MNC: %d",
+				cell_info.current_cell.id, cell_info.current_cell.tac,
+				cell_info.current_cell.mcc, cell_info.current_cell.mnc);
+		} else {
+			LOG_WRN("No current serving cell available");
+		}
+
+		if (cell_info.ncells_count || cell_info.gci_cells_count) {
+			LOG_INF("Performing multi-cell request with "
+				"%u neighbor cells and %u GCI cells",
+				cell_info.ncells_count, cell_info.gci_cells_count);
+		} else {
+			LOG_INF("Performing single-cell request");
+		}
+
+		loc_request.cell_info = &cell_info;
+		loc_request.wifi_info = wifi_info;
+		err = nrf_cloud_coap_location_get(&loc_request, &result);
+		(void)k_mutex_unlock(&cell_info_mutex);
+		if (err) {
+			LOG_ERR("Unable to get location: %d", err);
+			break;
+		}
+
+		if ((result.type >= 0) && (result.type < LOCATION_TYPE__INVALID)) {
+			type = location_types[result.type];
+		} else {
+			type = "unknown";
+		}
+		/* Process the returned location once it arrives */
+		gnss.pvt.lat = result.lat;
+		gnss.pvt.lon = result.lon;
+		gnss.pvt.accuracy = result.unc;
+		LOG_INF("Location: %s, %.12g, %.12g, %d", type,
+			result.lat, result.lon, result.unc);
+
+		request_cells = true;
+		break;
+	case 4:
+		printk("*** %d. Sending GNSS PVT ***********************\n", cur_test);
+		err = nrf_cloud_coap_location_send(&gnss);
+		if (err) {
+			LOG_ERR("Error sending GNSS PVT data: %d", err);
+		} else {
+			LOG_INF("PVT sent");
+		}
+		break;
+	case 5:
+		printk("*** %d. Getting A-GPS data *********************\n", cur_test);
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+		err = do_agps();
+#endif
+		break;
+	case 6:
+		printk("*** %d. Getting shadow delta *******************\n", cur_test);
+		buf[0] = '\0';
+		err = nrf_cloud_coap_shadow_get(buf, sizeof(buf) - 1, true);
+		if (err) {
+			LOG_ERR("Failed to request shadow delta: %d", err);
+		} else {
+			size_t len = strlen(buf);
+
+			LOG_INF("Delta: %s", len ? buf : "None");
+			if (len) {
+				err = nrf_cloud_coap_shadow_state_update(buf);
+				if (err) {
+					LOG_ERR("Failed to acknowledge delta: %d", err);
+				} else {
+					LOG_INF("Delta acknowledged");
+				}
+			}
+		}
+		break;
+	}
+
+	if (++cur_test > 6) {
+		cur_test = 1;
+	}
+	return err;
+}
+
+static void button_handler(uint32_t button_states, uint32_t has_changed)
+{
+	if (has_changed & button_states & BIT(BTN_NUM - 1)) {
+		LOG_INF("Button %d pressed", BTN_NUM);
+		button_pressed = true;
+	}
+}
+
+int main(void)
+{
+	int64_t next_msg_time;
+	int delta_ms = APP_COAP_SEND_INTERVAL_MS;
+	int err;
+	int i = 1;
+
+	LOG_INF("\n");
+	LOG_INF("The nRF Cloud CoAP client sample started\n");
+
+	do {
+		err = init();
+		if (err) {
+			/* NOTE: This is only meant for testing and debugging.
```

Comment:
I did this because Stig was having trouble getting it working, and he did not like that the sample just halted (it interrupted the ability to gather modem traces, apparently).

I am considering discarding most of this sample and rebuild it on top of MMSS's connection handling.  I'd rather not do that until the nrfcloud CoAP library and Tommi's location library changes are merged, so I can just use the location library.

Reply:
Changed it to do what your REST samples do instead.

Reply:
The issue with modem trace was because of the busy loop (`for (;;) {}`), and the application did never process the modem trace. This was unfortunate when not knowing the root cause of the issue.